foundation-models/openai/src/main/java/com/sap/ai/sdk/foundationmodels/openai/IterableStreamConverter.java

```diff
@@ -0,0 +1,113 @@
+package com.sap.ai.sdk.foundationmodels.openai;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Spliterator.NONNULL;
+import static java.util.Spliterator.ORDERED;
+
+import io.vavr.control.Try;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Spliterators;
+import java.util.concurrent.Callable;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.hc.core5.http.HttpEntity;
+
+/**
+ * Internal utility class to convert from a reading handler to {@link Iterable} and {@link Stream}.
+ *
+ * <p><strong>Note:</strong> All operations are sequential in nature. Thread safety is not
+ * guaranteed.
+ *
+ * @param <T> Iterated item type.
+ */
+@Slf4j
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+class IterableStreamConverter<T> implements Iterator<T> {
+  /** see {@link BufferedReader#DEFAULT_CHAR_BUFFER_SIZE} * */
+  static final int BUFFER_SIZE = 8192;
+
+  /** Read next entry for Stream or {@code null} when no further entry can be read. */
+  private final Callable<T> readHandler;
+
+  /** Close handler to be called when Stream terminated. */
+  private final Runnable stopHandler;
+
+  private boolean isDone = false;
+  private boolean isNextFetched = false;
+  private T next = null;
+
+  @SuppressWarnings("checkstyle:IllegalCatch")
+  @Override
+  public boolean hasNext() {
+    if (isDone) {
+      return false;
+    }
+    if (isNextFetched) {
+      return true;
+    }
+    try {
+      next = readHandler.call();
+      isNextFetched = true;
+      if (next == null) {
+        isDone = true;
+        stopHandler.run();
+      }
+    } catch (final Exception e) {
+      isDone = true;
+      stopHandler.run();
+      throw new IllegalStateException("Iterator stopped unexpectedly.", e);
```

Comment:
What can be thrown here?

Reply:
The exception caught here would be thrown by `readHandler.call()`.
In this case that would be from `BufferedReader#readLine`.
The `readLine` may throw `IOException – If an I/ O error occurs`.
Presumably this would happen when the nested `InputStream#read` would fail.

One possibility for an `InputStream` from `HttpEntity` to fail would be a **network interruption**.

Reply:
Then either do not throw anything or update the expection catching IOException and with a message like `Stream closed`. This is not an `IllegalStateException` if it can happen

Reply:
I agree, network issues will occur someday anyway. So let's improve error messages. https://github.com/SAP/ai-sdk-java/pull/49/commits/fb21544f9c622a38d08765066e545a585a34ae64